### DISCUSSION TITLE
Transactions between 2 accounts are imported multiple times creating duplicates

### DISCUSSION THREAD

**Discussion by SmokeySlim** [firefly-iii/firefly-iii#5906] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

If you have multiple accounts configured in FireFly-iii and you import transactions for those accounts any transfers between those accounts are imported twice. 

Looking at the debug info it appears that everything is identical except the create/update timestamp, which differs a few seconds.

Edit: it appears that the timestamps that are different are from the import and not the transaction itself, meaning the transaction information from the 2 accounts are entirely the same.

### Expected behaviour

If a transaction between 2 accounts is exactly the same except the timestamp differs with a few seconds it should be caught as a duplicate and not imported. The current behaviour imports duplicates causing issues.

Edit: it appears the timestamps are from the import and not from the bank. Duplicate transactions should be detected and not be imported.

### Debug information

Transaction A: 
{"id":2485,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"title":null,"transaction_journals":[{"id":2494,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"transaction_type_id":3,"transaction_group_id":2485,"bill_id":null,"transaction_currency_id":1,"description":"(no description)","date":"2022-02-26T23:00:00.000000Z","interest_date":null,"book_date":null,"process_date":null,"order":0,"tag_count":0,"completed":true,"transactions":[{"id":4987,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","deleted_at":null,"reconciled":false,"account_id":500,"transaction_journal_id":2494,"description":null,"transaction_currency_id":1,"amount":"-1450.000000000000","foreign_amount":null,"foreign_currency_id":null,"identifier":0,"account":{"id":500,"created_at":"2022-03-18T08:24:31.000000Z","updated_at":"2022-03-18T08:31:36.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"account_type_id":3,"name":"Account A","virtual_balance":"0.000000000000","iban":"IBAN A","active":true,"order":1},"budgets":[],"categories":[]},{"id":4988,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","deleted_at":null,"reconciled":false,"account_id":505,"transaction_journal_id":2494,"description":null,"transaction_currency_id":1,"amount":"1450.000000000000","foreign_amount":null,"foreign_currency_id":null,"identifier":0,"account":{"id":505,"created_at":"2022-03-18T08:30:28.000000Z","updated_at":"2022-03-18T08:31:36.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"account_type_id":3,"name":"Account B","virtual_balance":null,"iban":"IBAN B","active":true,"order":3},"budgets":[],"categories":[]}],"meta":[{"id":9827,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","transaction_journal_id":2494,"name":"payment_date","data":"2022-02-27 00:00:00","hash":"193d273a8da684c8bdf39cb4a9d7c821c63899daaa83c015a674db7ef1dc615c","deleted_at":null},{"id":9828,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","transaction_journal_id":2494,"name":"import_hash_v2","data":"286d31c388b5f029ef4cd677cc902c9cc5fd80d29f23d946fa60fbe0e747d037","hash":"c8f8281b7ef6ce3343bb95296c3671d59d060f784afced056e3b5cf31c722733","deleted_at":null},{"id":9829,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","transaction_journal_id":2494,"name":"external_id","data":"6b68bf65-6cf0-5dca-a079-5981c90ddc80","hash":"55739978057f22d4f2c61b71097a0b4f0243b05ddf25c2e8cf7376248a0bc929","deleted_at":null},{"id":9830,"created_at":"2022-03-18T08:44:02.000000Z","updated_at":"2022-03-18T08:44:02.000000Z","transaction_journal_id":2494,"name":"original_source","data":"ff3-v5.6.16|api-v1.5.5","hash":"08f51c7d6a812538833376e780f541a003d16d14f12bdcc22a6606529e462585","deleted_at":null}],"tags":[{"id":88,"created_at":"2022-03-18T08:43:54.000000Z","updated_at":"2022-03-18T08:43:54.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"tag":"Data Import on 2022-03-18 @ 09:43","tagMode":"nothing","date":"2022-03-17T23:00:00.000000Z","description":"","pivot":{"transaction_journal_id":2494,"tag_id":88}}],"categories":[],"budgets":[],"notes":[],"locations":[],"attachments":[],"links":[],"piggy_bank_events":[]}]}

Transaction B:
{"id":2558,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"title":null,"transaction_journals":[{"id":2567,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"transaction_type_id":3,"transaction_group_id":2558,"bill_id":null,"transaction_currency_id":1,"description":"(no description)","date":"2022-02-26T23:00:00.000000Z","interest_date":null,"book_date":null,"process_date":null,"order":0,"tag_count":0,"completed":true,"transactions":[{"id":5133,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","deleted_at":null,"reconciled":false,"account_id":500,"transaction_journal_id":2567,"description":null,"transaction_currency_id":1,"amount":"-1450.000000000000","foreign_amount":null,"foreign_currency_id":null,"identifier":0,"account":{"id":500,"created_at":"2022-03-18T08:24:31.000000Z","updated_at":"2022-03-18T08:31:36.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"account_type_id":3,"name":"Account A","virtual_balance":"0.000000000000","iban":"IBAN A","active":true,"order":1},"budgets":[],"categories":[]},{"id":5134,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","deleted_at":null,"reconciled":false,"account_id":505,"transaction_journal_id":2567,"description":null,"transaction_currency_id":1,"amount":"1450.000000000000","foreign_amount":null,"foreign_currency_id":null,"identifier":0,"account":{"id":505,"created_at":"2022-03-18T08:30:28.000000Z","updated_at":"2022-03-18T08:31:36.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"account_type_id":3,"name":"Account B","virtual_balance":null,"iban":"IBAN B","active":true,"order":3},"budgets":[],"categories":[]}],"meta":[{"id":10119,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","transaction_journal_id":2567,"name":"payment_date","data":"2022-02-27 00:00:00","hash":"193d273a8da684c8bdf39cb4a9d7c821c63899daaa83c015a674db7ef1dc615c","deleted_at":null},{"id":10120,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","transaction_journal_id":2567,"name":"import_hash_v2","data":"92469f573d4d53587bee8a95cae33748ea2c5a49b54016bbf2f4173345a3ff22","hash":"9eab0ffd06d82738f8d2c813ce2f9430b4a7f26433ee2dc836ec69e705c5ca01","deleted_at":null},{"id":10121,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","transaction_journal_id":2567,"name":"external_id","data":"4da085cb-25b6-51b6-ae54-664e7124548c","hash":"0ef1d26e377176f2cb1ea4b20c0913373b9b5c29df992f899f0aea2fe3e3d062","deleted_at":null},{"id":10122,"created_at":"2022-03-18T08:44:41.000000Z","updated_at":"2022-03-18T08:44:41.000000Z","transaction_journal_id":2567,"name":"original_source","data":"ff3-v5.6.16|api-v1.5.5","hash":"08f51c7d6a812538833376e780f541a003d16d14f12bdcc22a6606529e462585","deleted_at":null}],"tags":[{"id":88,"created_at":"2022-03-18T08:43:54.000000Z","updated_at":"2022-03-18T08:43:54.000000Z","deleted_at":null,"user_id":2,"user_group_id":null,"tag":"Data Import on 2022-03-18 @ 09:43","tagMode":"nothing","date":"2022-03-17T23:00:00.000000Z","description":"","pivot":{"transaction_journal_id":2567,"tag_id":88}}],"categories":[],"budgets":[],"notes":[],"locations":[],"attachments":[],"links":[],"piggy_bank_events":[]}]}

### Steps to reproduce

1. Create 2 or more accounts
2. Create one or more transactions between those accounts
3. Import transactions from both accounts

**Comment by JC5** [comment, marked as answer]
> If a transaction between 2 accounts is exactly the same except the timestamp differs with a few seconds it should be caught as a duplicate and not imported. 


Sorry, if the data is different between two transactions, it’s not a duplicate. I won’t be able to add exceptions.

**Reply by SmokeySlim** [reply]
> > If a transaction between 2 accounts is exactly the same except the timestamp differs with a few seconds it should be caught as a duplicate and not imported.
> 
> Sorry, if the data is different between two transactions, it’s not a duplicate. I won’t be able to add exceptions.

As mentioned in the edit, the timestamp isn't the problem. Please check the debug info - there's no difference its the same transaction.

**Reply by JC5** [reply]
Sorry, I’m on my phone. But the resulting transactions could be the same, they can be changed by rules and edits etc. So that’s not something I’m particularly concerned about. 

I’m curious to see what was submitted to Firefly III originally. If you have the logs, that would be helpful.

**Reply by JC5** [reply]
According to a quick check, the external ID is different between A and B.

**Reply by SmokeySlim** [reply]
I'm hoping you'll find what you need in these excerpts: 

[2022-03-18 09:33:50] local.DEBUG: Create Balance from array  
[2022-03-18 09:33:50] local.DEBUG: Downloaded transaction with ID "140a4db8-6f58-5117-9899-4965bf18f75c"  
[2022-03-18 09:33:50] local.DEBUG: Nordigen transaction from array {"entryReference":"1085","bookingDate":"2022-02-27","valueDate":"2022-02-27","transactionAmount":{"amount":"1450.00","currency":"EUR"},"creditorAccount":{"iban":"IBAN B","currency":"EUR"},"debtorName":"Account A","debtorAccount":{"iban":"IBAN A"},"key":"booked"} 

[2022-03-18 09:33:53] local.DEBUG: Create Balance from array  
[2022-03-18 09:33:53] local.DEBUG: Downloaded transaction with ID "c1d1a6ee-53f6-5bca-9857-1608c164103b"
[2022-03-18 09:33:53] local.DEBUG: Nordigen transaction from array {"entryReference":"6943","bookingDate":"2022-02-27","valueDate":"2022-02-27","transactionAmount":{"amount":"-1450.00","currency":"EUR"},"creditorName":"Account B","creditorAccount":{"iban":"IBAN B"},"debtorAccount":{"iban":"IBAN A","currency":"EUR"},"key":"booked"} 

[2022-03-18 09:34:31] local.DEBUG: Now submitting transaction 13/184  
[2022-03-18 09:34:31] local.DEBUG: Duplicate detection method is "classic", so this method is skipped (return true).  
[2022-03-18 09:34:31] local.DEBUG: Transaction #13 is unique.  
[2022-03-18 09:34:31] local.DEBUG: Going to map data for this line.  
[2022-03-18 09:34:31] local.DEBUG: Submitting to Firefly III: {"apply_rules":true,"error_if_duplicate_hash":true,"transactions":[{"type":"transfer","date":"2022-02-27","amount":"1450.00","description":"(no description)","payment_date":"2022-02-27","order":0,"currency_code":"EUR","tags":[],"category_name":null,"category_id":null,"notes":"","external_id":"fee92035-331c-507c-86ff-6f4703a1472d","internal_reference":"","destination_id":6,"source_id":1,"source_name":"Account A"}]} 

[2022-03-18 09:36:04] local.DEBUG: Now submitting transaction 85/184  
[2022-03-18 09:36:04] local.DEBUG: Duplicate detection method is "classic", so this method is skipped (return true).  
[2022-03-18 09:36:04] local.DEBUG: Transaction #85 is unique.  
[2022-03-18 09:36:04] local.DEBUG: Going to map data for this line.  
[2022-03-18 09:36:04] local.DEBUG: Submitting to Firefly III: {"apply_rules":true,"error_if_duplicate_hash":true,"transactions":[{"type":"transfer","date":"2022-02-27","amount":"1450.0000000000000000","description":"(no description)","payment_date":"2022-02-27","order":0,"currency_code":"EUR","tags":[],"category_name":null,"category_id":null,"notes":"","external_id":"650312dc-1626-596c-bca5-a83ff72bdcf3","internal_reference":"","source_id":1,"destination_id":6,"destination_name":"Account B"}]} 

That the external ID is different would make sense to me as the transaction is a transfer between 2 accounts that are in FireFly III. Both accounts have a record of that transfer with the bank. Account A gives 1450 to Account B. Both accounts should have a record of that with each holding their own unique ID of that transfer. 

If you need any other information please let me know.

**Reply by JC5** [reply]
Since the ID's are different the duplicate detector will not trigger. I don't know what else to tell you.

**Reply by SmokeySlim** [reply]
Well I would suggest that when there is a transaction being made between 2 known firefly-iii managed accounts to check for a similar transaction without comparing the ID's in order to solve that.

**Reply by JC5** [reply]
Firefly III used to have such a check but it's impossible to catch all the exceptions. The question, "when is a transaction similar?" is impossible to answer.

I gave up when I had a guy who complained that when he bought two Snicker bars in the same minute (two transactions), Firefly III would mark as a duplicate when it wasn't.

If your bank says it's two transactions (two different ID's) Firefly III will treat them as such.

**Reply by SmokeySlim** [reply]
I was hoping there'd be more accurate information given by the bank regarding the transaction. Unfortunately there isn't much more than the day the transaction was made. If they'd have hhmmss.ffff i would've had hope for a solution however you're right its not possible without introducing a myriad of problems.

Thanks for looking into this though!

**Reply by W0153R** [reply]
I just started with Firefly and I absolutely love it, great work!
My solution to this is the following:
- Account A has IBAN/BIC/etc
- Account B has _**no**_ IBAN/BIC/etc
- A rule that remove any withdrawals from account A to account B

When both accounts have account numbers, they will get recognized, very clever! Although it'll interfere when you're importing both accounts. That's the reason why, in this solution, account B doesn't have account numbers.

When you import account A, any transfers that would otherwise automatically get recognized, will get automatically removed instead.
When you import account B, it will recognize that it's coming from account A and will get automatically added to account A.
